Sample client code for .NET Framework 4.6.2 and Newtonsoft.Json 4.0.1

Shane32 opened the pull request:
See:
- https://github.com/graphql-dotnet/graphql-dotnet/issues/2748

sungam3r commented on src/ClientNet462/packages.config:
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Newtonsoft.Json" version="4.0.1" targetFramework="net462" />
```
4.0.1 ? So old?

Shane32 replied:
The sample was requested for this specific version.  The code works with no changes with Newtonsoft.Json 13.0.1 also.

sungam3r replied:
I think that was a mistake. I see 12.0.3 in zipped sample. 4.0.1 is a version of `GraphQL.Client` 😉

Shane32 replied:
Well, no matter.  This sample works with basically any version of Newtonsoft.Json.

sungam3r replied:
OK

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,116 @@
+
+        private static HttpClient httpClient = new HttpClient();
+
+        static async Task<GraphQlResponse<TResponse>> CallGraphQlAsync<TResponse>(Uri endpoint, HttpMethod method, string query, object variables, CancellationToken cancellationToken)
```
```suggestion
        static async Task<GraphQLResponse<TResponse>> CallGraphQLAsync<TResponse>(Uri endpoint, HttpMethod method, string query, object variables, CancellationToken cancellationToken)
```

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,116 @@
+    internal class Program
+    {
+        static void Main(string[] args)
+        {
+            MainAsync(default).Wait();
+        }
```
```suggestion
        {
            using var cts = new CancellationTokenSource();
            Console.CancelKeyPress += (o, e) => cts.Cancel();
            MainAsync(cts.Token).Wait();
        }
```

To demostrate working cancellation.

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,116 @@
+            var result = await CallGraphQlAsync<MyResponseData>(
+                new Uri("https://graphql-weather-api.herokuapp.com/"),
+                HttpMethod.Post,
+                "query ($city: String!) { getCityByName(name: $city) { name country } }",
```
By the way, `getCityByName` is bad naming for GraphQL data field.

Shane32 replied:
I only selected this endpoint and query because is an actual public API and does return data, even if a bit pointless.

sungam3r replied:
I understand.

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,116 @@
+            }
+        }
+
+        public class GraphQlErrorLocation
```
```suggestion
        public class GraphQLErrorLocation
```

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,116 @@
+            public int Column { get; set; }
+        }
+
+        public class GraphQlError
```
```suggestion
        public class GraphQLError
```

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,116 @@
+            public List<object> Path { get; set; } //either int or string
+        }
+
+        public class GraphQlResponse<TResponse>
```
```suggestion
        public class GraphQLResponse<TResponse>
```

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,116 @@
+            public TResponse Data { get; set; }
+        }
+
+        private static string SerializeGraphQlCall(string query, object variables)
```
```suggestion
        private static string SerializeGraphQLCall(string query, object variables)
```

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,116 @@
+            return sb.ToString();
+        }
+
+        private static GraphQlResponse<TResponse> DeserializeGraphQlCall<TResponse>(string response)
```
```suggestion
        private static GraphQLResponse<TResponse> DeserializeGraphQLCall<TResponse>(string response)
```

sungam3r commented:
👍 Well, this is actuall server examples but I think it will be even better to have example for client app. Please add new line about new example into the readme table.

Shane32 commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,148 @@
+                    ConsoleCancelEventHandler handler = (o, e) =>
+                    {
+                        Console.WriteLine("Cancelling...");
+                        cts.CancelAfter(0);
+                        e.Cancel = true;
```
`cts.Cancel();` executes synchronously, throwing a `TaskCancellationException` on the other thread, which hits the finally block to disassociate the `CancelKeyPress` handler, which results in a deadlock!!!!  ugh!  Solved with `CancelAfter(0);`.  Also, without `e.Cancel = true;`, the process immediately terminates, which appears as a hanged application when debugging.

sungam3r replied:
WOW!

Shane32 replied:
Yeah, sometimes multithreading really comes back to bite you! - lol

sungam3r replied:
Well,  `Console.CancelKeyPress -= handler;` actually is not needed for such run-and-terminate console app. Does deadlock occur without it?

Shane32 replied:
no, but if a user was to press Ctrl-C afterwards it would throw an ObjectDisposedException when accessing the CancellationTokenSource.  I figured as a matter of 'good form' it should show proper removal of the handler prior to disposing of the cts.  Of course I could have either (a) not disposed of the cts at all, or (b) put the other console code within the using block.

Shane32 replied:
I just figured that it's always best to have samples with proper tear-down code.

Shane32 replied:
Now realizing that `Cancel` is synchronous, it probably should properly be `CancelAfter` anyway in this sample, in case of any other locks put in place by `Console`.

sungam3r replied:
What mean "Cancel is synchronous"?

Shane32 replied:
`Cancel()` is a blocking operation that executes all registered cancellation notification delegates.  The docs say it can throw an `AggregateException` if any delegate throws an exception, so this is known behavior.  Apparently in this case the registered cancellation delegate within the `HttpClient` is also executing a blocking operation when throwing exceptions on pending `Task`s.  The `finally` clause then creates the deadlock attempting to edit an event handler while `Console` is executing the handlers (probably placing a lock on the list while iterating through it).

See: https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.cancel?view=net-6.0#System_Threading_CancellationTokenSource_Cancel

sungam3r replied:
> registered cancellation notification delegates

These are those using `cts.Token.Register(() => throw new Exception("aaa"));`. Example has no such registrations (I don't see) so nothing bad can happen. I have a few CLI tools (Core, not NET) that use such approach with `CancellationTokenSource` and `ConsoleKeyPress` handler. They work without problems, the cancellation works, there are no dedlocks. I do not know... maybe it's NET differs from Core. But for the purposes of the example, it is not important.

Shane32 commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,148 @@
+                                city = "detroit",
+                            },
+                            cts.Token);
+                        if (result.Errors?.Count > 0)
+                        {
+                            Console.WriteLine($"GraphQL returned errors:\n{string.Join("\n", result.Errors.Select(x => $"  - {x.Message}"))}");
+                        }
```
Figured something basic to show returned errors wouldn't hurt.

sungam3r replied:
👍

Shane32 commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,148 @@
+            httpRequestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            using (var response = await httpClient.SendAsync(httpRequestMessage, cancellationToken))
+            {
+                //if (response.IsSuccessStatusCode)
+                if (response?.Content.Headers.ContentType?.MediaType == "application/json")
```
This particular endpoint has a BadRequest status code for bad graphql queries.  In any case, this seemed more appropriate for any endpoint, although often perhaps the other condition would be better.

sungam3r replied:
bad graphql queries - these having `errors` field ?

Shane32 replied:
I guess.  For testing I tried changing the query to have a bad field name.  The endpoint returned a proper GraphQL response with the `errors` field and so on, but with a BadRequest HTTP status code.

sungam3r replied:
Worth to mention that such GraphQL server does not match the spec.

sungam3r replied:
Ah, then OK. It's OK to return non-200 if execution stage didn't begin.

Shane32 commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,148 @@
+                //if (response.IsSuccessStatusCode)
+                if (response?.Content.Headers.ContentType?.MediaType == "application/json")
+                {
+                    var responseString = await response.Content.ReadAsStringAsync(); //cancellationToken supported for .NET 5/6
```
Don't know why, but none of the methods on `HttpContent` prior to .NET 5 support a cancellation token here.  However, the sample still works because most of the waiting period is during `SendAsync`, which properly supports the cancellation token.

sungam3r replied:
Does not matter

Shane32 commented:
> Please change to 12.0.3 or 13 - any "actual" version

Newtonsoft.Json?  4.0.1 is an "actual" version.  It is not deprecated or anything.

sungam3r commented:
OK

Shane32 commented on README.md:
```diff
@@ -10,6 +10,7 @@ Examples for [GraphQL.NET](https://github.com/graphql-dotnet/graphql-dotnet).
 | AspNetWebApi     | .NET Framework 4.6.1 Web API example |
 | AzureFunctions   | Using GraphQL in Azure Functions |
 | StarWars         | StarWars demo schema used in all examples |
+| ClientNet461     | Example GraphQL client written in .NET 4.6.1 using Newtonsoft.Json 4.0.1 |
```
```suggestion
| ClientNet461     | Example GraphQL client written in .NET 4.6.2 using Newtonsoft.Json 4.0.1 |
```

Shane32 commented on README.md:
```diff
@@ -10,6 +10,7 @@ Examples for [GraphQL.NET](https://github.com/graphql-dotnet/graphql-dotnet).
 | AspNetWebApi     | .NET Framework 4.6.1 Web API example |
 | AzureFunctions   | Using GraphQL in Azure Functions |
 | StarWars         | StarWars demo schema used in all examples |
+| ClientNet461     | Example GraphQL client written in .NET 4.6.2 using Newtonsoft.Json 4.0.1 |
```
```suggestion
| ClientNet462     | Example GraphQL client written in .NET 4.6.2 using Newtonsoft.Json 4.0.1 |
```

sungam3r commented on src/ClientNet462/Program.cs:
```diff
@@ -0,0 +1,148 @@
+        {
+            try
+            {
+                using (var cts = new CancellationTokenSource())
```
```suggestion
                using var cts = new CancellationTokenSource();
```

Saves 1 indent for example.

Shane32 replied:
Not supported in .NET 4.6.2

Shane32 replied:
(Or more specifically, the version of C# required for that syntax isn't supported by .NET 4.6.2, or something like that.)

sungam3r replied:
🤦

Shane32 commented:
@sungam3r I added a bunch of comments and refactored Main again so that the sample code wasn't mixed in with all the wire-up.  (`Main` was getting too complex to understand easily.)  Do you want to review again?

sungam3r commented:
OK to merge.
